Use `config_for` to hold omniauth config hash

mjankowski opened the pull request:
This uses the Rails `config_for` feature to pull in a yaml file and convert it to hash settings on `Rails.configuration`.

Multiple motivations here:

- In general I think we should consolidate the loading of env vars as early as possible in the load process. Read the values once from config, initializers, etc at loadtime, and then access them via config objects throughout the code. Will surface missing values sooner, and add some level of DRY/readability improvement in my opinion to access these via config objects in app code
- In this particular case, we were also manually building up some hash objects, which is a great match for yaml

If we like this general direction, there are other areas that would benefit from the same approach.

Side notes:

- The `omniauth_only?` method is repeteated a few times with same definition. I'm guessing there's a reason for this, but may be worth reviewing/consolidating that.
- I did a little tidying up with `fetch` in the yaml file to read the values in, and believe I preserved the original defaults, but I'll do a little more self-review on this specific part to make sure it's preserved
- Definitely open to naming suggestions here on both the file name (and thus top level config name), top-level keys, sub-keys, etc...

mjankowski committed:
Use `config_for` to hold omniauth config hash

mjankowski committed:
Convert scope to symbol array outside yaml read

ClearlyClaire reviewed (changes requested):
From a readability point of view, that's a net improvement, and I also agree with setting the values once and early.

However, I am not sure injecting our custom `config` properties is really clean or future-safe, and there is at least one typo in a constant name.

I'm also worried setting `nil` values to hash keys is going to behave differently than not setting the hash key at all for all of these environment variables, but I have not actually checked how OmniAuth and its providers actually treat those different hashes.

renchap commented:
There is a Rails feature to avoid adding our options to the global `config` (and risk issues in the future): [`config.x`](https://guides.rubyonrails.org/configuring.html#config-x)

mjankowski committed:
Update config/omniauth.yml

Co-authored-by: Claire <claire.github-309c@sitedethib.com>

mjankowski committed:
Read oidc scope as yaml symbol array

mjankowski commented:
> There is a Rails feature to avoid adding our options to the global `config` (and risk issues in the future): [`config.x`](https://guides.rubyonrails.org/configuring.html#config-x)

Yeah, I think the guidance from before the `config_for` feature existed was something like "Just use top-level `config.your_value` if you have one value, and use `config.x.your_values.key...` when you have either multiple related values from a namespace, or a big hash or whatever.

With the introduction of `config_for`, the guidance (on that same page) - https://guides.rubyonrails.org/configuring.html#custom-configuration - seems to be the same, but to add "use `config_for` to load whole configuration files" (and since the format it supports is yaml files, these are almost always going to be nested hashes).

I guess if we want to be conservative here, we could read from yaml with config_for and ALSO put it under `x` namepsace -- something like `config.x.omniauth = ...` in application.rb and `Rails.configuration.x.omniauth...` when we read values?

mjankowski commented:
> From a readability point of view, that's a net improvement, and I also agree with setting the values once and early.

Excellent - this one is probably one of the more complex ones because of how many values there are. I'll look for other areas that we can move env access to early-setting early-failing config vars.

> However, I am not sure injecting our custom `config` properties is really clean or future-safe, and

Addressed in another comment - but I think given the feature is framework supported its probably ok ... but we could also find a more conservative namespace to use if we want to be extra careful.

> there is at least one typo in a constant name.

Yup - I'm going to do another self-review pass through this to check the values. This also seems sort of absurd, but I could see doing another PR in advance of this one which expands our "does setting this env var correctly lead to the right config options being passed in" sort of specs first, and then rebasing this one to ensure we preserve the behaviour.

> I'm also worried setting `nil` values to hash keys is going to behave differently than not setting the hash key at all for all of these environment variables, but I have not actually checked how OmniAuth and its providers actually treat those different hashes.

Good point, that angle had not occurred to me, but it's certainly possible that "this hash value is nil" vs "this hash key does not even exist" COULD be treated differently. I'll try to look into this and either confirm it doesn't matter, or update code to preserve the prior behavior.

ClearlyClaire commented on config/application.rb:
```diff
@@ -100,6 +100,9 @@ class Application < Rails::Application
       require 'mastodon/redis_config'
     end
 
+    # Load config/omniauth.yml settings
+    config.omniauth = config_for(:omniauth)
```
I meant the use of the top-level Rails `configuration` namespace. `omniauth` is probably safe indeed, though.

ClearlyClaire commented:
I can confirm this changes the behavior of default values.

Indeed, taking the example of `OIDC_UID_FIELD`, the behavior stays the same when the environment variable is defined, but when omitted, `nil` is used instead of the library's default of `'sub'`, as evidenced by checking: `Devise.omniauth_configs[:openid_connect].strategy['uid_field']`

mjankowski commented:
Indeed. Will restore this behavior, maybe add some config loading specs first.

Before that though:

- do we like this general approach of wrapping these with config_for to move env loading earlier?
- thoughts on naming, namespace? Re: https://github.com/mastodon/mastodon/pull/30534

ClearlyClaire commented:
> do we like this general approach of wrapping these with config_for to move env loading earlier?

At least for this part, yes, if the defaults issue can be handled in a graceful way.

> thoughts on naming, namespace? Re: https://github.com/mastodon/mastodon/pull/30534

Hm, I don't have a strong opinion on this, but I think we should use `config.x.omniauth`, as we're clearly in the case of a nested hash?

mjankowski committed:
Move to `x` namespace

mjankowski commented:
Rebased this and moved to `x` namespace. It occurs to me that it may be way easier to review this if we do it in two steps:

- First PR just moves the top level "enabled check" vars to a `config_for` setup, and modifies the various checks around the app
- Follow-up PR (or maybe one-per-auth-type) to move the line-by-line env var settings over as well

Going to see if I can easily pull out a diff branch from this that just does the first, and try that as separate PR.
